src/core/service.c

```diff
@@ -2659,6 +2662,67 @@ static void service_enter_reload_by_notify(Service *s) {
                 log_unit_warning(UNIT(s), "Failed to schedule propagation of reload, ignoring: %s", bus_error_message(&error, r));
 }
 
+static bool service_should_reload_extensions(Service *s) {
+        int r;
+
+        assert(s);
+
+        /* Only support this for notify-reload service types. */
+        if (s->type != SERVICE_NOTIFY_RELOAD)
+                return false;
+
+        /* TODO: Add support for user services, which can use
+         * ExtensionDirectories= + notify-reload. For now, skip for user
+         * services. */
+        if (UNIT(s)->manager->runtime_scope != RUNTIME_SCOPE_SYSTEM) {
+                log_unit_debug(UNIT(s), "Not reloading extensions for user services.");
+                return false;
+        }
+
+        r = exec_context_has_vpicked_extensions(&s->exec_context);
+        if (r < 0) {
+                log_unit_warning_errno(UNIT(s), r, "Failed to determine if service should reload extensions, assuming false: %m");
+                return false;
+        }
+        return r > 0;
+}
+
+static int service_reload_extensions(Service *s) {
+        /* TODO: do this asynchronously */
+        _cleanup_free_ char *propagate_dir = NULL;
+
+        assert(s);
+
+        /* TODO: remove after adding support for user services */
+        assert(UNIT(s)->manager->runtime_scope == RUNTIME_SCOPE_SYSTEM);
```

Comment:
it's fine not to support per-user reloads for now. But this should not be triggering an assert if people try. 

please move this check into service_should_reload_extensions() and make it graceful

Reply:
This might be like the 4th time you make the same comment  :-)
The graceful check is already there in service_should_reload_extensions() so this is 
just there to ensure a programming error doesn't sneak in later https://github.com/systemd/systemd/pull/33995/commits/77eb9bf10d69eac49d6264a978e208bf76fddbec#diff-88a01402a6feb6c8a13f32a718325c7fdbdd4ca47de52b147ff75eaace8e1903R2677

Reply:
oh, you already have it there! then please just drop the assert here

Reply:
But the assert is useful to ensure things are not accidentally regressing, any reason to drop it?

Reply:
look at the code. you first do the assert() check, and then you immediately after do the graceful check. So you will always be dead by the time where you can gracefully fail...

at the very least the order of this assert and the following check should be reversed. But frankly, just drop the assert(), there's no point in checking the same thing twice in lines basically following each other immediately.

Reply:
Ah yes, that is indeed something to fix, missed that